### PR TITLE
feat: ledger integration

### PR DESCRIPTION
**Pull request by stanleyyconsensys**
## **Description**
_Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:_
_1. What is the reason for the change?_
_2. What is the improvement/solution?_

## **Manual testing steps**

_1. Step1:_
_2. Step2:_
_3. ..._

## **Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the before and after of your change._

### **Before**

_[screenshot]_

### **After**

_[screenshot]_

## **Related issues**

_Fixes #???_

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained:
  - [ ] What problem this PR is solving.
  - [ ] How this problem was solved.
  - [ ] How reviewers can test my changes.
- [ ] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/util/address/index.js [hunk 1]
```diff
@@ -167,6 +211,43 @@ export function isQRHardwareAccount(address) {
...
+export function getLabelTextByAddress(address) {
+  if (!address) return null;
+  if (isHardwareAccount(address, [HardwareDeviceNames.ledger]))
+    return 'accounts.ledger';
+  if (isHardwareAccount(address, [KeyringTypes.qr]))
+    return 'transaction.hardware';
+  if (isImportedAccount(address)) return 'accounts.imported';
+
+  return null;
+}
```
Comment (gantunesr) [inline app/util/address/index.js:250]: There's some inconsistencies regarding Ledger and QR wallets. Would be good to double check that.

File: app/util/hardwareWallet/signatureUtils.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    //TODO: add error handling
+    throw new Error(`Keyring not found for address ${messageParams.from}`);
+  }
+
+  const navPromise = navMethodFactory.get(keyring.type as HardwareDeviceNames);
+
+  if (navPromise === undefined) {
+    //TODO: add error handling
+    throw new Error(`No nav type for address type ${keyring.type}`);
+  }
```
Comment (gantunesr) [inline app/util/hardwareWallet/signatureUtils.ts:35]: Improve error message

File: app/util/test/initial-background-state.json [hunk 3]
```diff
@@ -183,4 +183,4 @@
     "unapprovedPersonalMsgCount": 0,
     "unapprovedTypedMessagesCount": 0
   }
-}
+}
```
Comment (gantunesr) [inline app/util/test/initial-background-state.json:186]: Revert this change

File: ios/MetaMask/Info.plist [hunk 4]
```diff
@@ -43,6 +43,10 @@
 	<string>$(CURRENT_PROJECT_VERSION)</string>
 	<key>ITSAppUsesNonExemptEncryption</key>
 	<false/>
+	<key>LSApplicationQueriesSchemes</key>
+	<array>
+		<string>twitter</string>
+	</array>
```
Comment (gantunesr) [inline ios/MetaMask/Info.plist:49]: Check why this was added

File: package.json [hunk 5]
```diff
@@ -558,4 +563,4 @@
       "ts-node>@swc/core": false
     }
   }
-}
+}
```
Comment (gantunesr) [inline package.json:566]: Revert this change

File: app/components/Views/LedgerConnect/hooks/useBluetoothPermissions.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        bluetoothAllowed =
+          connectPermissionStatus === RESULTS.GRANTED &&
+          scanPermissionStatus === RESULTS.GRANTED;
+      } else {
+        bluetoothAllowed = bluetoothPermissionStatus === RESULTS.GRANTED;
+      }
+
+      if (bluetoothAllowed) {
+        setHasBluetoothPermissions(true);
+        //Edge case, when `View setting` click, checkPermissions will trigger, and setBluetoothPermissionError(undefined), but when screen move away to android setting page,
```
Comment (gantunesr) [inline app/components/Views/LedgerConnect/hooks/useBluetoothPermissions.ts:61]: nit
```suggestion
        // Edge case, when `View setting` click, checkPermissions will trigger, and setBluetoothPermissionError(undefined), but when screen move away to android setting page,
```

File: app/constants/navigation/Routes.ts [hunk 7]
```diff
@@ -18,6 +18,19 @@ const Routes = {
...
   },
+  HW: {
+    CONNECT: 'ConnectHardwareWalletFlow',
+    SELECT_DEVICE: 'SelectHardwareWallet',
+    CONNECT_QR_DEVICE: 'ConnectQRHardwareFlow',
+    CONNECT_LEDGER: 'LedgerConnectFlow',
+    LEDGER_ACCOUNT: 'LedgerAccountInfo',
+  },
+  LEDGER_CONNECT_FLOW: {
+    ID: 'LedgerConnectFlow',
```
Comment (gantunesr) [inline app/constants/navigation/Routes.ts:29]: Duplicated values

File: app/components/UI/SignatureRequest/Root/Root.tsx [hunk 8]
```diff
@@ -32,13 +33,16 @@ const Root = ({
   const navigation = useNavigation();
   const { colors } = useTheme();
   const [showExpandedMessage, setShowExpandedMessage] = useState(false);
+  const visability = useSelector(
```
Comment (gantunesr) [inline app/components/UI/SignatureRequest/Root/Root.tsx:36]:
```suggestion
  const visibility = useSelector(
```

File: app/components/UI/SignatureRequest/Root/Root.tsx [hunk 9]
```diff
@@ -32,13 +33,16 @@ const Root = ({
...
+    (reduxState: any) => reduxState.modals.signMessageModalVisible,
+  );
 
   const toggleExpandedMessage = () =>
     setShowExpandedMessage(!showExpandedMessage);
 
   const currentPageMeta = messageParams?.meta;
 
-  if (!messageParams || !currentPageMeta || !approvalType) {
+  if (!messageParams || !currentPageMeta || !approvalType || !visability) {
```
Comment (gantunesr) [inline app/components/UI/SignatureRequest/Root/Root.tsx:45]:
```suggestion
  if (!messageParams || !currentPageMeta || !approvalType || !visibility) {
```

File: app/components/UI/SignatureRequest/Root/Root.tsx (hunk 8 above)
Comment (gantunesr) [reply app/components/UI/SignatureRequest/Root/Root.tsx:36]: Consider changing the variable name

File: android/.project [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
-			<name>org.eclipse.buildship.core.gradleprojectbuilder</name>
-			<arguments>
-			</arguments>
-		</buildCommand>
-	</buildSpec>
-	<natures>
-		<nature>org.eclipse.buildship.core.gradleprojectnature</nature>
-	</natures>
+        <name>MetaMask</name>
+        <comment>Project MetaMask created by Buildship.</comment>
```
Comment (gantunesr) [inline android/.project:4]: Revert this change

File: app/component-library/components/Avatars/Avatar/foundation/AvatarBase/AvatarBase.test.tsx [hunk 11]
```diff
@@ -26,4 +26,19 @@ describe('AvatarBase', () => {
...
+    const wrapper = shallow(
+      <AvatarBase>
+        <Image
+          source={{ uri: toDataUrl(DUMMY_IMAGE_DATA) }}
+          style={blockieStyles}
+        />
+      </AvatarBase>,
+    );
+    expect(wrapper).toMatchSnapshot();
+  });
```
Comment (gantunesr) [inline app/component-library/components/Avatars/Avatar/foundation/AvatarBase/AvatarBase.test.tsx:43]: Looks like this change should not be part of this PR

File: app/util/address/index.js (hunk 1 above)
Comment (stanleyyconsensys) [reply app/util/address/index.js]: as prev comment
Ledger keyring logic should not implement via keyring controller
hence, we havent touch the keyring controller and propose to add it via mobile directly
however, qr code still maintain in keyring controller for some reason
hence we adding a new type to map to the ledger keyring

File: app/component-library/components/Avatars/Avatar/foundation/AvatarBase/AvatarBase.test.tsx (hunk 11 above)
Comment (stanleyyconsensys) [reply app/component-library/components/Avatars/Avatar/foundation/AvatarBase/AvatarBase.test.tsx]: resolved

File: android/.project (hunk 10 above)
Comment (stanleyyconsensys) [reply android/.project]: resolved

File: app/components/UI/SignatureRequest/Root/Root.tsx (hunk 9 above)
Comment (stanleyyconsensys) [reply app/components/UI/SignatureRequest/Root/Root.tsx]: resolved

File: app/components/UI/SignatureRequest/Root/Root.tsx (hunk 8 above)
Comment (stanleyyconsensys) [reply app/components/UI/SignatureRequest/Root/Root.tsx]: resolved

File: app/constants/navigation/Routes.ts (hunk 7 above)
Comment (stanleyyconsensys) [reply app/constants/navigation/Routes.ts]: resolved by removing LEDGER_CONNECT_FLOW, instead merging into HW

File: app/components/Views/LedgerConnect/hooks/useBluetoothPermissions.ts (hunk 6 above)
Comment (stanleyyconsensys) [reply app/components/Views/LedgerConnect/hooks/useBluetoothPermissions.ts]: resolved

File: app/core/BackgroundBridge/BackgroundBridge.js [hunk 12]
```diff
@@ -155,10 +159,14 @@ export class BackgroundBridge extends EventEmitter {
...
-      params: false,
-    });
+    try {
+      this.sendNotification({
+        method: NOTIFICATION_NAMES.unlockStateChanged,
+        params: false,
+      });
+    } catch (e) {
+      // Do nothing
+    }
```
[the remark this answers is not in the document]
Comment (stanleyyconsensys) [reply app/core/BackgroundBridge/BackgroundBridge.js]: resolved

File: app/components/Nav/Main/RootRPCMethodsUI.js [hunk 13]
```diff
@@ -203,7 +209,31 @@ const RootRPCMethodsUI = (props) => {
           },
         );
         await KeyringController.resetQRKeyringState();
-        Engine.acceptPendingApproval(transactionMeta.id);
+
+        // eslint-disable-next-line no-console
+        console.log(transactionMeta);
```
[the remark this answers is not in the document]
Comment (stanleyyconsensys) [reply app/components/Nav/Main/RootRPCMethodsUI.js]: resolved

File: app/components/Nav/App/index.js [hunk 14]
```diff
@@ -574,6 +579,29 @@ const App = ({ userLoggedIn }) => {
     </Stack.Navigator>
   );
 
+  // eslint-disable-next-line react/prop-types
```
[the remark this answers is not in the document]
Comment (stanleyyconsensys) [reply app/components/Nav/App/index.js]: not necessary, removed

File: app/component-library/components/Avatars/Avatar/foundation/AvatarBase/__snapshots__/AvatarBase.test.tsx.snap
Comment (stanleyyconsensys) [inline app/component-library/components/Avatars/Avatar/foundation/AvatarBase/__snapshots__/AvatarBase.test.tsx.snap:1]: @dawnseeker8 can we remove this?

File: ios/MetaMask.xcodeproj/project.pbxproj
Comment (stanleyyconsensys) [inline ios/MetaMask.xcodeproj/project.pbxproj:1]: why we need to update this file?

File: ios/MetaMask/Info.plist (hunk 4 above)
Comment (stanleyyconsensys) [reply ios/MetaMask/Info.plist:49]: i think this is duplicated, and should be remove

File: ios/Podfile.lock [hunk 15]
```diff
@@ -930,6 +952,6 @@ SPEC CHECKSUMS:
   Yoga: ba09b6b11e6139e3df8229238aa794205ca6a02a
   YogaKit: f782866e155069a2cca2517aafea43200b01fd5a
 
-PODFILE CHECKSUM: 8d3adf4eeb9c38c94e164bf39d459809e3147413
+PODFILE CHECKSUM: a75a7b4687a558d94615247d0e8a266e714632e5
 
-COCOAPODS: 1.12.1
```
Comment (stanleyyconsensys) [inline ios/Podfile.lock:935]: why we need to change the version

**Comment by gantunesr** [comment]
Closed in favor of https://github.com/MetaMask/metamask-mobile/pull/7757
